packages/react-dom/src/__tests__/ReactUpdates-test.js

```diff
@@ -1734,4 +1739,57 @@ describe('ReactUpdates', () => {
       });
     }).toThrow('Maximum update depth exceeded');
   });
+
+  fit('prevents infinite update loop between a component with synchronous updates in useEffect and a component with synchronous updates in render', async () => {
+    // Ignore flushSync warning
+    spyOnDev(console, 'error');
+
+    function A({stateRef}) {
+      const [a, updateA] = React.useState(false);
+      stateRef.current.updateA = updateA;
+      stateRef.current.a = a;
+      React.useEffect(() => {
+        const {updateB} = stateRef.current;
+        ReactDOM.flushSync(() => {
+          updateB(true);
+        })
+      });
+      return <div onClick={() => console.log('A')}>A</div>;
+    }
+    
+    function B({stateRef}) {
+      const {updateA, a} = stateRef.current;
+      const [b, updateB] = React.useState(false);
+      stateRef.current.updateB = updateB;
+      React.useEffect(() => {
+        updateB(false);
+      });
+      if (updateA) {
+        updateA(!a);
```

Comment:
This is accompanied by a warning, right?

Reply:
There is a `Maximum update depth exceeded` exception on this call, it is thrown as a recoverable error by `onRecoverableError`.